tikv-client/src/main/java/com/pingcap/tikv/predicates/TiKVScanAnalyzer.java

```diff
@@ -297,47 +297,116 @@ TiKVScanPlan buildIndexScan(
       Range<TypedKey> r = ir.getRange();
-
+      BigDecimal max = new BigDecimal(Long.MAX_VALUE);
       if (!r.hasLowerBound()) {
         // -INF
-        startKey = RowKey.createMin(id);
+        signedStartKey = RowKey.createMin(id);
       } else {
         // Comparison with null should be filtered since it yields unknown always
-        startKey = RowKey.toRowKey(id, r.lowerEndpoint());
-        if (r.lowerBoundType().equals(BoundType.OPEN)) {
-          startKey = startKey.next();
+        /**
+         * {@link com.pingcap.tikv.types.IntegerType#decodeNotNull(int, CodecDataInput)}
+         */
+        Object obj = r.lowerEndpoint().getValue();
+        if (obj instanceof Long) {
+          signedStartKey = RowKey.toRowKey(id, new IntHandle((long) obj));
+          if (r.lowerBoundType().equals(BoundType.OPEN)) {
+            signedStartKey = signedStartKey.next();
+          }
+        } else if (obj instanceof BigDecimal) {
+          isUnsignedLong = true;
+          BigDecimal bigDecimal = (BigDecimal) obj;
+          // since range is left-closed and right-open interval,
+          // we have to use next to exclusive Long.MAX_VALUE.
+          signedStartKey = RowKey.toRowKey(id, new IntHandle(max.longValue())).nextPrefix();
+          unsignedStartKey = RowKey.toRowKey(id, new IntHandle(max.longValue() + 1));
+          if (bigDecimal.compareTo(max) > 0) {
+            unsignedStartKey = RowKey.toRowKey(id, new IntHandle(bigDecimal.longValue()));
+            if (r.lowerBoundType().equals(BoundType.OPEN)) {
+              unsignedStartKey = unsignedStartKey.nextPrefix();
+            }
+          } else {
+            signedStartKey = RowKey.toRowKey(id, new IntHandle(bigDecimal.longValue()));
+            if (r.lowerBoundType().equals(BoundType.OPEN)) {
+              signedStartKey = signedStartKey.nextPrefix();
+            }
+          }
         }
       }
 
       if (!r.hasUpperBound()) {
         // INF
-        endKey = RowKey.createBeyondMax(id);
+        if (!isUnsignedLong) {
+          signedEndKey = RowKey.createBeyondMax(id);
+        } else {
+          // since range is left-closed and right-open interval,
+          // we have to use next to inclusive Long.MAX_VALUE.
+          signedEndKey = RowKey.toRowKey(id, new IntHandle(max.longValue())).nextPrefix();
+          unsignedEndKey = RowKey.createBeyondMax(id);
+        }
       } else {
-        endKey = RowKey.toRowKey(id, r.upperEndpoint());
-        if (r.upperBoundType().equals(BoundType.CLOSED)) {
-          endKey = endKey.next();
+        /**
+         * {@link com.pingcap.tikv.types.IntegerType#decodeNotNull(int, CodecDataInput)}
+         */
+        Object obj = r.upperEndpoint().getValue();
+        if (obj instanceof Long) {
+          signedEndKey = RowKey.toRowKey(id, new IntHandle((long) obj));
+          if (r.upperBoundType().equals(BoundType.CLOSED)) {
+            signedEndKey = signedEndKey.nextPrefix();
```

Comment:
same as before

Reply:
change to when not unsigned long or other type of long will throw exception